processing/src/main/java/io/druid/segment/DimensionIndexer.java

```diff
@@ -0,0 +1,189 @@
+ * a value->ID dictionary as row values are seen.
+ *
+ * The DimensionIndexer is also responsible for implementing various value lookup operations,
+ * such as conversion between an encoded value and its full representation. It maintains knowledge of the
+ * mappings between encoded values and actual values.
+ *
+ * NOTE:
+ * When encoding is present, there are two relevant orderings for the encoded values.
+ *
+ * 1.) Ordering based on encoded value's order of ingestion
+ * 2.) Ordering based on converted actual value
+ *
+ * Suppose we have a new String dimension DimA, which sees the values "Hello", "World", and "Apple", in that order.
+ * This would correspond to dictionary encodings of "Hello"=0, "World"=1, and "Apple"=2, by the order
+ * in which these values were first seen during ingestion.
+ *
+ * However, some use cases require the encodings to be sorted by their associated actual values.
+ * In this example, that ordering would be "Apple"=0, "Hello"=1, "World"=2.
+ *
+ * @param <EncodedType> class of the encoded values
+ * @param <ActualType> class of the actual values
+ *
+ */
+public interface DimensionIndexer<EncodedType extends Comparable<EncodedType>, ActualType extends Comparable<ActualType>>
+{
+
+  /**
+   * Given a list of row values (for multivalued dimensions) or single row value, update any
+   * internal data structures with the ingested values and return the row values as a Comparable[]
+   * to be used as a TimeAndDims key.
+   *
+   * The values within the Comparable[] return value should be encoded if applicable, i.e. as instances of EncodedType.
+   *
+   * NOTE: This function can change the internal state of the DimensionIndexer.
+   *
+   * @param dimValues Row value(s) to process
+   *
+   * @return A new DimensionIndexer object that synchronizes on the provided lock.
+   */
+  public Object processRowValsToIndexKey(Object dimValues);
+
+
+  /**
+   * Some rows may not contain a particular dimension. Suppose the schema has DimA and DimB.
+   * If an event {"dimB"="abcd"} appears, dimA's DimensionIndexer will not see that null value for DimA.
+   * This function is called by IncrementalIndexAdapter to inform the DimensionIndexer of such unseen null values.
+   */
+  public void addNullLookup();
+
+
+  /**
+   * Given an encoded value, return its full representation.
+   *
+   * Using the example in the class description:
+   *   getActualValue(0, true) would return "Apple"
+   *   getActualValue(0, false) would return "Hello"
+   *
+   * @param intermediateValue The encoded value to be converted
+   * @param idSorted true, if the encoded value was sorted by actual value
+   *                 false, if the encoded value was ordered by time of ingestion
+   *
+   * @return The actual value associated with the encoded intermediateValue
+   *
+   */
+  public ActualType getActualValue(EncodedType intermediateValue, boolean idSorted);
```

Comment:
I think it would be better to encapsulate if you are interacting with the sorted versus unsorted id space in a separate interface/class and have a method that returns that.  Maintaining a boolean in order to ensure that you are referencing the correct thing is error prone.